### DISCUSSION TITLE
How to reuse the last result of a rule in the same rule ?

### DISCUSSION THREAD

**Discussion by BlancoMY** [lf-edge/ekuiper#2277] [Q&A]
Hello, 

How can I efficiently reuse (if it is possible) the last result of a computation within the same computation, such as in the equation A = A + B?  my usecase is i'm reading data (B) from a PLC using EgdeX Foundry and i want to calculate A = A + B where A is the result of the rule (the first execution where A does not exist can be ignored or set to a default value)  i tried to republish the result to the EdgeX bus (**edgex/rules-events** topic) but i could't make it work. 
I could implement a work around using two rules C = A+B , D = C+B but i was wondering if there is a way to have it in one rule.
 
Additionally, Is there any full example of using windows in graph-rule, such as calculating an average with tumbling window? (I found an example with sql query but nothing with graph)

Thank you

**Comment by ngjaying** [comment]
@EscanorUt We have some stateful functions called analytic functions. One of it is lag to get the last value of any expression. Please check if it can meet your requirements. doc here: https://ekuiper.org/docs/en/latest/sqls/functions/analytic_functions.html#lag

**Reply by BlancoMY** [reply]
@ngjaying Thank you for your quick answer. I have already tried with **lag** and **latest** but it didn't work.

`SELECT avg(Temperature) + lag(avg_temp) AS avg_temp FROM source1 WHERE Temperature != nil GROUP BY TUMBLINGWINDOW(ss, 20) HAVING avg(Temperature) > 0`

I get this kind of warnings:

`time="2023-09-26 07:55:00" level=warning msg="No readings are processed for the event, so ignore it." file="edgex/edgex_source.go:198" rule=rule1
time="2023-09-26 07:55:00" level=warning msg="No readings are processed for the event, so ignore it." file="edgex/edgex_source.go:198" rule=rule1Graph`

I think the only way is to use two rules to achieve my usecase. 

Thank you

**Comment by ngjaying** [comment]
Regarding the window in graph API, here is an example and I'll update the doc to include it.
```json
{
  "id": "ruleGraphWindow",
  "name": "Rule that uses window",
  "graph": {
    "nodes": {
      "demo": {
        "type": "source",
        "nodeType": "mqtt",
        "props": {
          "datasource": "devices/win/messages",
          "timestamp": "ts"
        }
      },
      "roundfunc": {
        "type": "operator",
        "nodeType": "function",
        "props": {
          "expr": "round(temperature) as temp"
        }
      },
      "window": {
        "type": "operator",
        "nodeType": "window",
        "props": {
          "type": "tumblingwindow",
          "unit": "ms",
          "size": 600
        }
      },
      "avgfunc": {
        "type": "operator",
        "nodeType": "aggfunc",
        "props": {
          "expr": "avg(temp) as avg_temp"
        }
      },
      "pick" : {
        "type": "operator",
        "nodeType": "pick",
        "props": {
          "fields": ["avg_temp", "humidity", "window_end() as win"]
        }
      },
      "mqtt": {
        "type": "sink",
        "nodeType": "mqtt",
        "props": {
          "server": "tcp://syno.home:1883",
          "topic": "result/graphWindow",
          "sendSingle": true
        }
      },
      "log": {
        "type": "sink",
        "nodeType": "log",
        "props": {}
      }
    },
    "topo": {
      "sources": ["demo"],
      "edges": {
        "demo": ["roundfunc"],
        "roundfunc": ["window"],
        "window": ["avgfunc"],
        "avgfunc": ["pick"],
        "pick": ["mqtt", "log"]
      }
    }
  },
  "options": {
    "isEventTime": true,
    "lateTolerance" : 0
  }
}
```

**Reply by BlancoMY** [reply]
This is so helpfull, thank you
